speasy/core/http.py

```diff
@@ -1,27 +1,111 @@
 from speasy import __version__
 import platform
 import requests
+import socket
 from requests.utils import quote as _quote
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+from urllib.request import Request, urlopen, HTTPError, URLError
 from time import sleep
 import logging
 
 log = logging.getLogger(__name__)
 
 USER_AGENT = f'Speasy/{__version__} {platform.uname()} (SciQLop project)'
 
+DEFAULT_TIMEOUT = 60  # seconds
+
+DEFAULT_DELAY = 5  # seconds
+
+DEFAULT_RETRY_COUNT = 5
+
+STATUS_FORCE_LIST = [500, 502, 503, 504]  # Note: Specific treatment for 429 error code (see below)
+
+
+class TimeoutHTTPAdapter(HTTPAdapter):
+    def __init__(self, *args, **kwargs):
+        self.timeout = DEFAULT_TIMEOUT
+        if "timeout" in kwargs:
+            self.timeout = kwargs["timeout"]
+            del kwargs["timeout"]
+        super().__init__(*args, **kwargs)
+
+    def send(self, request, **kwargs):
+        timeout = kwargs.get("timeout")
+        if timeout is None:
+            kwargs["timeout"] = self.timeout
+        return super().send(request, **kwargs)
+
+
 def quote(*args, **kwargs):
     return _quote(*args, **kwargs)
 
-def get(url, headers: dict = None, params: dict = None):
+
+def apply_delay(headers: dict = None):
+    delay = DEFAULT_DELAY
+    try:
+        if headers and ('Retry-After' in headers):
+            delay = float(headers['Retry-After'])
+    except ValueError:
+        pass
+    log.debug(f"Will sleep for {delay} seconds")
+    sleep(delay)
+
+
+def get(url, headers: dict = None, params: dict = None, timeout: int = DEFAULT_TIMEOUT, head_only: bool = False):
     headers = {} if headers is None else headers
     headers['User-Agent'] = USER_AGENT
-    resp = requests.get(url, headers=headers, params=params)
-    while resp.status_code in [429, 503]:
-        try:
-            delay = float(resp.headers['Retry-After'])
-        except ValueError:
-            delay = 5
-        log.debug(f"Got {resp.status_code} response, will sleep for {delay} seconds")
-        sleep(delay)
-        resp = requests.get(url, headers=headers, params=params)
+    # cf. https://findwork.dev/blog/advanced-usage-python-requests-timeouts-retries-hooks/
+    retry_strategy = Retry(
+        total=DEFAULT_RETRY_COUNT,
+        backoff_factor=1,
+        status_forcelist=STATUS_FORCE_LIST,
+        allowed_methods=["HEAD", "GET"]
+    )
+    adapter = TimeoutHTTPAdapter(max_retries=retry_strategy, timeout=timeout)
+    http = requests.Session()
+    http.mount("https://", adapter)
+    http.mount("http://", adapter)
+    while True:
+        if head_only:
+            resp = http.head(url, headers=headers, params=params)
+        else:
+            resp = http.get(url, headers=headers, params=params)
+        if resp.status_code == 429:  # Honor "Retry-After"
```

Comment:
We should still check for 503 too since CDAWeb uses it:
https://cdaweb.gsfc.nasa.gov/WebServices/REST/  (Get data endpoint)